src/VectorBundle.jl

```diff
@@ -0,0 +1,372 @@
+
+"""
+    project_vector!(M::VectorBundleFibers, v, x, w)
+
+Project vector `w` from the vector space of type `VS` at point `x`
+and save the result to `v`.
+"""
+function project_vector!(M::VectorBundleFibers, v, x, w)
+    error("project_vector! not implemented for vector space manifold $(typeof(M)), vector space of type $(typeof(VS)), output vector of type $(typeof(v)) and input vector at point $(typeof(x)) with type of w $(typeof(w)).")
+end
+
+function project_vector!(M::VectorBundleFibers{<:TangentSpaceType}, v, x, w)
+    project_tangent!(M.M, v, x, w)
+    return v
+end
+
+"""
+    VectorBundle(M::Manifold, type::VectorSpaceType)
+
+Vector bundle on manifold `M` of type `type`.
+"""
+struct VectorBundle{TVS<:VectorSpaceType, TM<:Manifold} <: Manifold
+    type::TVS
+    M::TM
+    VS::VectorBundleFibers{TVS, TM}
+end
+
+function VectorBundle(VS::TVS, M::TM) where{TVS<:VectorSpaceType, TM<:Manifold}
+    return VectorBundle{TVS, TM}(VS, M, VectorBundleFibers(VS, M))
+end
+
+function representation_size(M::VectorBundle)
+    len_manifold = prod(representation_size(M.M))
+    len_vs = prod(representation_size(M.VS))
+    return (len_manifold + len_vs,)
+end
+
+TangentBundleFibers{M} = VectorBundleFibers{TangentSpaceType,M}
+TangentBundleFibers(M::Manifold) = VectorBundleFibers(TangentSpace, M)
+
+CotangentBundleFibers{M} = VectorBundleFibers{CotangentSpaceType,M}
+CotangentBundleFibers(M::Manifold) = VectorBundleFibers(CotangentSpace, M)
+
+TangentBundle{M} = VectorBundle{TangentSpaceType,M}
+TangentBundle(M::Manifold) = VectorBundle(TangentSpace, M)
+
+CotangentBundle{M} = VectorBundle{CotangentSpaceType,M}
+CotangentBundle(M::Manifold) = VectorBundle(CotangentSpace, M)
+
+"""
+    bundle_projection(M::VectorBundle, x::ProductRepr)
+
+Projection of point `x` from the bundle `M` to the base manifold.
+Returns the point on the base manifold `M.M` at which the vector part
+of `x` is attached.
+"""
+function bundle_projection(M::VectorBundle, x)
+    return submanifold_component(x, 1)
+end
+
+function isapprox(M::VectorBundle, x, y; kwargs...)
+    return isapprox(M.M, x.parts[1], y.parts[1]; kwargs...) &&
+        isapprox(x.parts[2], y.parts[2]; kwargs...)
+end
+
+function isapprox(M::VectorBundle, x, v, w; kwargs...)
+    return isapprox(M.M, v.parts[1], w.parts[1]; kwargs...) &&
+        isapprox(M.M, x.parts[1], v.parts[2], w.parts[2]; kwargs...)
+end
+
+manifold_dimension(M::VectorBundle) = manifold_dimension(M.M) + vector_space_dimension(M.VS)
+
+function inner(M::VectorBundle, x, v, w)
+    return inner(M.M, x.parts[1], v.parts[1], w.parts[1]) +
+           inner(M.VS, x.parts[2], v.parts[2], w.parts[2])
+end
+
+function distance(M::VectorBundle, x, y)
+    dist_man = distance(M.M, x.parts[1], y.parts[1])
+    vy_x = vector_transport(M.M, y.parts[1], y.parts[2], x.parts[1])
+    dist_vec = vector_distance(M.VS, x.parts[1], x.parts[2], vy_x)
+
+    return sqrt(dist_man^2 + dist_vec^2)
+end
+
+function exp!(M::VectorBundle, y, x, v)
```

Comment:
Oh this is cool! I think we're missing `exp` and `log`? Also, would love to see this documented in both `exp` and `vector_transport`, for the users who want geodesic flow.

Reply:
Default versions of `exp` and `log` seem to work OK (they are tested). What exactly do you want documented for these functions? I think `exp` for vector bundles could be just described in general documentation for vector bundles.

Reply:
> Default versions of `exp` and `log` seem to work OK (they are tested).

Oh, of course you're right. I was thinking `VectorBundle` wasn't a `Manifold`, but it is.

> What exactly do you want documented for these functions?

TBH, I'm looking at the code for `log`, and I don't get what it's doing. Same with `exp`. Would be nice to have it in the documentation.